File: webapp/updates.py
"""
Module to handle /updates API calls.
"""

from cache import REPO_LABEL, REPO_BASEARCH, REPO_RELEASEVER, REPO_URL, PKG_SUMMARY_ID, PKG_DESC_ID, PKG_EVR_ID, \
    ERRATA_CVE, ERRATA_TYPE
from common.webapp_utils import none2empty, filter_package_list
from common.rpm import parse_rpm_name, join_rpm_name

SECURITY_ERRATA_TYPE = 'security'


class UpdatesAPI:
    """ Main /updates API class."""
    def __init__(self, db_cache):
        self.db_cache = db_cache      # DB dump in memory, stored like a dict

    def _process_repositories(self, data, response):
        # Read list of repositories
        repo_list = data.get('repository_list', None)
        if repo_list is not None:
            repo_ids = []
            for label in repo_list:
                repo_id = self.db_cache.repolabel2ids.get(label, None)
                if repo_id:
                    repo_ids.extend(repo_id)
            response['repository_list'] = repo_list
        else:
            repo_ids = self.db_cache.repo_detail.keys()

        # Filter out repositories of different releasever
        releasever = data.get('releasever', None)
        if releasever is not None:
            repo_ids = [oid for oid in repo_ids
                        if self.db_cache.repo_detail[oid][REPO_RELEASEVER] == releasever
                        or (self.db_cache.repo_detail[oid][REPO_RELEASEVER] is None
                            and releasever in self.db_cache.repo_detail[oid][REPO_URL])]
            response['releasever'] = releasever

        # Filter out repositories of different basearch
        basearch = data.get('basearch', None)
        if basearch is not None:
            repo_ids = [oid for oid in repo_ids
                        if self.db_cache.repo_detail[oid][REPO_BASEARCH] == basearch
                        or (self.db_cache.repo_detail[oid][REPO_BASEARCH] is None
                            and basearch in self.db_cache.repo_detail[oid][REPO_URL])]
            response['basearch'] = basearch

        return set(repo_ids)

    def _process_input_packages(self, data, response):
        """Parse input NEVRAs and filter out unknown (or without updates) package names."""
        latest_only = data.get("latest_only", False)
        packages_to_process = filter_package_list(data.get('package_list', None), latest_only)
        filtered_packages_to_process = {}
        if packages_to_process is not None:
            for pkg in packages_to_process:
                response['update_list'][pkg] = {}
                name, epoch, ver, rel, arch = parse_rpm_name(pkg, default_epoch='0')
                if name in self.db_cache.packagename2id:
                    if self.db_cache.packagename2id[name] in self.db_cache.updates_index:
                        filtered_packages_to_process[pkg] = {'parsed_nevra': (name, epoch, ver, rel, arch)}
        return filtered_packages_to_process

    def _get_valid_releasevers(self, original_package_repo_ids):
        valid_releasevers = set()
        for original_package_repo_id in original_package_repo_ids:
            valid_releasevers.add(self.db_cache.repo_detail[original_package_repo_id][REPO_RELEASEVER])
        return valid_releasevers

    def _get_repositories(self, update_pkg_id, errata_ids, available_repo_ids, valid_releasevers):
        repo_ids = set()
        errata_repo_ids = set()
        for errata_id in errata_ids:
            errata_repo_ids.update(self.db_cache.errataid2repoids.get(errata_id, []))

        res_repos = set(self.db_cache.pkgid2repoids.get(update_pkg_id, []))
        res_repos.intersection_update(available_repo_ids, errata_repo_ids)

        for repo_id in res_repos:
            repo_detail = self.db_cache.repo_detail[repo_id]
            if repo_detail[REPO_RELEASEVER] in valid_releasevers:
                repo_ids.add(repo_id)

        return repo_ids

    def _build_nevra(self, update_pkg_id):
        name_id, evr_id, arch_id, _, _, _ = self.db_cache.package_details[update_pkg_id]
        name = self.db_cache.id2packagename[name_id]
        epoch, ver, rel = self.db_cache.id2evr[evr_id]
        arch = self.db_cache.id2arch[arch_id]
        return join_rpm_name(name, epoch, ver, rel, arch)

    def _process_updates(self, packages_to_process, api_version, available_repo_ids,
                         response, module_ids, security_only, optimistic_updates):
        # pylint: disable=too-many-branches
        for pkg, pkg_dict in packages_to_process.items():
            name, epoch, ver, rel, arch = pkg_dict['parsed_nevra']
            name_id = self.db_cache.packagename2id[name]
            evr_id = self.db_cache.evr2id.get((epoch, ver, rel), None)
            arch_id = self.db_cache.arch2id.get(arch, None)
            current_evr_indexes = []

            if evr_id is not None:
                current_evr_indexes = self.db_cache.updates_index[name_id].get(evr_id, [])
            elif optimistic_updates:
                # Try to find an evra, which can be used as an anchor point
                for idx, pkg_id in enumerate(self.db_cache.updates[name_id]):
                    evr = self.db_cache.id2evr.get(self.db_cache.package_details[PKG_EVR_ID], None)
                    if evr > (epoch, ver, rel):
                        current_evr_indexes = list(range(idx, len(self.db_cache.updates[name_id])))

            # Package with given NEVRA not found in cache/DB
            if not current_evr_indexes:
                continue

            current_nevra_pkg_id = None
            for current_evr_index in current_evr_indexes:
                pkg_id = self.db_cache.updates[name_id][current_evr_index]
                current_nevra_arch_id = self.db_cache.package_details[pkg_id][2]
                if current_nevra_arch_id == arch_id:
                    current_nevra_pkg_id = pkg_id
                    break

            # Package with given NEVRA not found in cache/DB
            if not current_nevra_pkg_id:
                continue

            if api_version == 1:
                sum_id = self.db_cache.package_details[current_nevra_pkg_id][PKG_SUMMARY_ID]
                response['update_list'][pkg]['summary'] = self.db_cache.strings.get(sum_id, None)

                desc_id = self.db_cache.package_details[current_nevra_pkg_id][PKG_DESC_ID]
                response['update_list'][pkg]['description'] = self.db_cache.strings.get(desc_id, None)

            response['update_list'][pkg]['available_updates'] = []

            # No updates found for given NEVRA
            last_version_pkg_id = self.db_cache.updates[name_id][-1]
            if last_version_pkg_id == current_nevra_pkg_id:
                continue

            # Get associated product IDs
            original_package_repo_ids = set()
            original_package_repo_ids.update(self.db_cache.pkgid2repoids.get(current_nevra_pkg_id, []))
            valid_releasevers = self._get_valid_releasevers(original_package_repo_ids)

            # Get candidate package IDs
            update_pkg_ids = self.db_cache.updates[name_id][current_evr_indexes[-1] + 1:]

            for update_pkg_id in update_pkg_ids:
                # Filter out packages without errata
                if update_pkg_id not in self.db_cache.pkgid2errataids:
                    continue

                # Filter arch compatibility
                updated_nevra_arch_id = self.db_cache.package_details[update_pkg_id][2]
                if (updated_nevra_arch_id != arch_id
                        and updated_nevra_arch_id not in self.db_cache.arch_compat[arch_id]):
                    continue

                errata_ids = self.db_cache.pkgid2errataids.get(update_pkg_id, set())
                nevra = self._build_nevra(update_pkg_id)
                for errata_id in errata_ids:
                    errata_name = self.db_cache.errataid2name[errata_id]
                    # Filter out non-security updates
                    if security_only and not (
                            self.db_cache.errata_detail[errata_name][ERRATA_TYPE] == SECURITY_ERRATA_TYPE or \
                            self.db_cache.errata_detail[errata_name][ERRATA_CVE]):
                        continue
                    if ((update_pkg_id, errata_id) in self.db_cache.pkgerrata2module and not \
                            self.db_cache.pkgerrata2module[(update_pkg_id, errata_id)].intersection(module_ids)):
                        continue
                    repo_ids = self._get_repositories(update_pkg_id, [errata_id], available_repo_ids,
                                                      valid_releasevers)
                    for repo_id in repo_ids:
                        repo_details = self.db_cache.repo_detail[repo_id]
                        response['update_list'][pkg]['available_updates'].append({
                            'package': nevra,
                            'erratum': errata_name,
                            'repository': repo_details[REPO_LABEL],
                            'basearch': none2empty(repo_details[REPO_BASEARCH]),
                            'releasever': none2empty(repo_details[REPO_RELEASEVER])
                        })

    def process_list(self, api_version, data):
        """
        This method is looking for updates of a package, including name of package to update to,
        associated erratum and repository this erratum is from.

        :param data: input json, must contain package_list to find updates for them

        :returns: json with updates_list as a list of dictionaries
                  {'package': <p_name>, 'erratum': <e_name>, 'repository': <r_label>}
        """
        # pylint: disable=too-many-branches
        response = {
            'update_list': {},
        }

        # Get list of valid repository IDs based on input paramaters
        available_repo_ids = self._process_repositories(data, response)
        modules_list = data.get('modules_list', [])
        module_info = [(x['module_name'], x['module_stream']) for x in modules_list]
        module_ids = set()
        for module in module_info:
            if module in self.db_cache.modulename2id:
                module_ids.update(self.db_cache.modulename2id[module])
        if 'modules_list' in data:
            response['modules_list'] = modules_list

        # Backward compatibility of older APIs
        if api_version < 3:
            security_only = True
        else:
            security_only = data.get("security_only", False)

        optimistic_updates = data.get("optimistic_updates", False)

        # Return empty update list in case of empty input package list
        packages_to_process = self._process_input_packages(data, response)

        if not packages_to_process:
            return response

        # Process updated packages, errata and fill the response
        self._process_updates(packages_to_process, api_version,
                              available_repo_ids, response, module_ids,
                              security_only, optimistic_updates)

        return response

File: webapp-next/calc/updates/updates.go
package updates

import (
	"app/cache"
	"app/utils"
	"github.com/pkg/errors"
	"sort"
	"strings"
)

type Request struct {
	Packages []string             `json:"package_list"`
	RepoList []string             `json:"repository_list"`
	Modules  []cache.ModuleStream `json:"modules_list"`

	Releasever *string `json:"releasever"`
	BaseArch   *string `json:"basearch"`
}

func (r Request) Validate() (int, error) {
	if r.Packages == nil {
		return 400, errors.New("'package_list' is a required property")
	}
	for _, m := range r.Modules {
		if len(m.Module) == 0 {
			return 400, errors.New("'module_name' is a required property")
		}
		if len(m.Stream) == 0 {
			return 400, errors.New("'module_stream' is a required property")
		}
	}
	return 200, nil
}

type Update struct {
	Package    string `json:"package"`
	Erratum    string `json:"erratum"`
	Repository string `json:"repository"`
	Basearch   string `json:"basearch"`
	Releasever string `json:"releasever"`
}

type NameUpdateDetail struct {
	// TODO: Force VMaaS to either contain available_updates or skip the object entirely
	AvailableUpdates *[]Update `json:"available_updates,omitempty"`

	Description string `json:"description,omitempty"`
	Summary     string `json:"summary,omitempty"`
}

type Response struct {
	UpdateList map[string]NameUpdateDetail `json:"update_list"`
	RepoList   []string                    `json:"repository_list,omitempty"`
	ModuleList *[]cache.ModuleStream       `json:"modules_list,omitempty"`
	Releasever *string                     `json:"releasever,omitempty"`
	BaseArch   *string                     `json:"basearch,omitempty"`
}

func nil2empty(s *string) string {
	if s != nil {
		return *s
	}
	return ""
}
func nil2debug(s *string) string {
	if s != nil {
		return *s
	}
	return "<NIL>"
}

func ProcessRepositories(c *cache.Cache, req Request, resp *Response) (map[cache.RepoID]bool, error) {
	repos := make(map[cache.RepoID]bool)

	if len(req.RepoList) > 0 {
		for _, r := range req.RepoList {
			if ids, has := c.RepoLabel2Ids[r]; has {
				for _, l := range ids {
					repos[l] = true
				}
			}
		}
		resp.RepoList = req.RepoList
	} else {
		for id := range c.RepoDetails {
			repos[id] = true
		}
	}

	// Slow but works, iterate over every item, ANDing it with another condition
	// WARNING ! do string comparison instead of ptr comparison
	if req.Releasever != nil {
		for r := range repos {
			det := c.RepoDetails[r]
			repos[r] = repos[r] &&
				(det.ReleaseVer == nil && strings.Contains(det.Url, *req.Releasever)) ||
				(det.ReleaseVer != nil && *det.ReleaseVer == *req.Releasever)
		}
		resp.Releasever = req.Releasever
	}

	if req.BaseArch != nil {
		for r := range repos {
			det := c.RepoDetails[r]
			repos[r] = repos[r] &&
				(det.BaseArch == nil && strings.Contains(det.Url, *req.BaseArch)) ||
				(det.BaseArch != nil && *det.BaseArch == *req.BaseArch)
		}
		resp.BaseArch = req.BaseArch
	}

	for r, keep := range repos {
		if !keep {
			delete(repos, r)
		}
	}
	return repos, nil
}

func ProcessInputPackages(cache *cache.Cache, data Request, resp *Response) map[string]utils.Nevra {
	pkgs := map[string]utils.Nevra{}

	for _, p := range data.Packages {
		nevra, err := utils.ParseNevra(p)
		if resp.UpdateList == nil {
			resp.UpdateList = map[string]NameUpdateDetail{}
		}
		resp.UpdateList[p] = NameUpdateDetail{}
		if err != nil {
			continue
		}
		if _, has := cache.Packagename2Id[nevra.Name]; !has {
			continue
		}
		pkgs[p] = *nevra
	}
	return pkgs
}

func getRelatedProducts(c *cache.Cache, repoids map[cache.RepoID]bool) map[int]bool {
	products := make(map[int]bool)
	for p := range repoids {
		products[c.RepoDetails[p].ProductId] = true
	}
	return products
}

func getReleaseVersions(c *cache.Cache, originalRepoIds map[cache.RepoID]bool) map[string]bool {
	releasevers := map[string]bool{}
	for rid := range originalRepoIds {
		rv := c.RepoDetails[rid].ReleaseVer
		releasevers[nil2empty(rv)] = true
	}
	return releasevers
}

func getRepositories(c *cache.Cache, updatePkgid cache.PkgID, productIds map[int]bool, relesevers map[string]bool, errataIds []cache.ErrataID, availableRepoIds map[cache.RepoID]bool) map[cache.RepoID]bool {
	repoids := map[cache.RepoID]bool{}
	errataRepoIds := map[cache.RepoID]bool{}

	for _, errataId := range errataIds {
		for _, rid := range c.ErrataId2RepoIds[errataId] {
			if availableRepoIds[rid] {
				errataRepoIds[rid] = true
			}
		}
	}

	for _, rid := range c.PkgId2RepoIds[updatePkgid] {
		if errataRepoIds[rid] {
			detail := c.RepoDetails[rid]
			if productIds[detail.ProductId] && relesevers[nil2empty(detail.ReleaseVer)] {
				repoids[rid] = true
			}
		}
	}
	return repoids
}

func buildNevra(c *cache.Cache, updatePkgId cache.PkgID) string {
	det := c.PackageDetails[updatePkgId]
	name := c.Id2Packagename[det.NameId]
	evr := c.Id2Evr[det.EvrId]
	arch := c.Id2Arch[det.ArchId]
	nevra := utils.Nevra{
		Name:    name,
		Epoch:   evr.Epoch,
		Version: evr.Version,
		Release: evr.Release,
		Arch:    arch,
	}
	return nevra.String()
}

func checkSecurityOnly(c *cache.Cache, securityOnly bool, errataId cache.ErrataID) bool {
	if !securityOnly {
		return true
	}
	errataName, has := c.ErrataId2Name[errataId]
	if !has {
		utils.Log("err", "errataid").Error("Missing entry in errataid2name")
	}
	if c.ErrataDetail[errataName].Type == "security" || len(c.ErrataDetail[errataName].CVEs) != 0 {
		return true
	}

	return false
}
func checkModules(c *cache.Cache, modules map[int]bool, updatePkgId cache.PkgID, errataId cache.ErrataID) bool {

	if len(modules) == 0 {
		return false
	}

	pkgErrata := cache.PkgErrata{
		PkgId:    int(updatePkgId),
		ErrataId: int(errataId),
	}
	mods, has := c.PkgErrata2Module[pkgErrata]
	if !has {
		// Not a module related errata-pkg pair
		return true
	}

	for _, m := range mods {
		if has := modules[m]; !has {
			return false
		}
	}
	return true
}

func ProcessUpdates(c *cache.Cache,
	pkgs map[string]utils.Nevra,
	repos map[cache.RepoID]bool,
	modules map[int]bool,
	response *Response,
	includeTexts,
	securityOnly,
	filterModules bool,
) error {
	for pkgString, pkg := range pkgs {
		evr := cache.Evr{
			Epoch:   pkg.Epoch,
			Version: pkg.Version,
			Release: pkg.Release,
		}

		nameId := c.Packagename2Id[pkg.Name]
		evrId := c.Evr2Id[evr]
		archId := c.Arch2Id[pkg.Arch]

		currentEvrIndexes := c.UpdatesIndex[nameId][evrId]
		if len(currentEvrIndexes) == 0 {
			continue
		}

		currentNevraPkgId := cache.PkgID(0)
		for _, idx := range currentEvrIndexes {
			pkgId := c.Updates[nameId][idx]
			// The package with same arch as searched from list specified in updatesIndex
			if c.PackageDetails[pkgId].ArchId == archId {
				currentNevraPkgId = pkgId
			}
		}

		if currentNevraPkgId == 0 {
			continue
		}

		response.UpdateList[pkgString] = NameUpdateDetail{
			AvailableUpdates: &[]Update{},
		}

		if includeTexts {
			updateList := response.UpdateList[pkgString]
			updateList.Summary = c.String[c.PackageDetails[currentNevraPkgId].SummaryId]
			updateList.Description = c.String[c.PackageDetails[currentNevraPkgId].DescriptionId]
			response.UpdateList[pkgString] = updateList
		}

		lastVersionPkgId := c.Updates[nameId][len(c.Updates[nameId])-1]
		if lastVersionPkgId == currentNevraPkgId {
			continue
		}

		originalPkgRepoids := map[cache.RepoID]bool{}
		for _, id := range c.PkgId2RepoIds[currentNevraPkgId] {
			originalPkgRepoids[id] = true
		}

		productIds := getRelatedProducts(c, originalPkgRepoids)
		validReleaseVers := getReleaseVersions(c, originalPkgRepoids)

		updatePkgIds := c.Updates[nameId][currentEvrIndexes[len(currentEvrIndexes)-1]+1:]

		for _, updatePkgId := range updatePkgIds {
			if _, has := c.PkgId2ErrataIds[updatePkgId]; !has {
				continue
			}
			updatedNevraArchId := c.PackageDetails[updatePkgId].ArchId

			if !c.ArchCompat[archId][updatedNevraArchId] {
				continue
			}

			errataIds := c.PkgId2ErrataIds[updatePkgId]
			pkgNevra := buildNevra(c, updatePkgId)

			for _, errataId := range errataIds {
				if !checkSecurityOnly(c, securityOnly, errataId) {
					continue
				}
				if filterModules && !checkModules(c, modules, updatePkgId, errataId) {
					continue
				}
				repoIds := getRepositories(c, updatePkgId, productIds, validReleaseVers, []cache.ErrataID{errataId}, repos)

				for r := range repoIds {
					detail := c.RepoDetails[r]
					updates := response.UpdateList[pkgString]
					tmp := append(*updates.AvailableUpdates, Update{
						Package:    pkgNevra,
						Erratum:    c.ErrataId2Name[errataId],
						Repository: detail.Label,

						Basearch:   nil2empty(detail.BaseArch),
						Releasever: nil2empty(detail.ReleaseVer),
					})
					updates.AvailableUpdates = &tmp
					response.UpdateList[pkgString] = updates
				}

			}
			slice := *response.UpdateList[pkgString].AvailableUpdates
			sort.Slice(slice, func(i, j int) bool {
				return slice[i].Package < slice[j].Package
			})
		}

	}
	return nil
}

func Updates(cache *cache.Cache, data Request, includeStrings, securityOnly bool) (Response, error) {
	response := Response{}
	pkgs := ProcessInputPackages(cache, data, &response)

	repos, err := ProcessRepositories(cache, data, &response)

	modules := map[int]bool{}
	var filterModules bool
	if data.Modules != nil {
		filterModules = true
		for _, m := range data.Modules {
			for _, mid := range cache.ModuleName2Ids[m] {
				modules[mid] = true
			}
		}
		response.ModuleList = &data.Modules
	}

	err = ProcessUpdates(cache, pkgs, repos, modules, &response, includeStrings, securityOnly, filterModules)
	return response, err
}
